src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/WSSecurityJan2004.cs

```diff
@@ -0,0 +1,579 @@
+            internal static readonly XmlDictionaryString ValueTypeAttribute = CoreWCF.XD.SecurityJan2004Dictionary.ValueType;
+
+            private string[] _valueTypeUris = null;
+
+            protected BinaryTokenEntry(string valueTypeUri)
+            {
+                _valueTypeUris = new string[1];
+                _valueTypeUris[0] = valueTypeUri;
+            }
+
+            protected BinaryTokenEntry(string[] valueTypeUris)
+            {
+                if (valueTypeUris == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(valueTypeUris));
+                }
+
+                _valueTypeUris = new string[valueTypeUris.GetLength(0)];
+                for (int i = 0; i < _valueTypeUris.GetLength(0); ++i)
+                {
+                    _valueTypeUris[i] = valueTypeUris[i];
+                }
+            }
+
+            protected override XmlDictionaryString LocalName { get { return ElementName; } }
+            protected override XmlDictionaryString NamespaceUri { get { return CoreWCF.XD.SecurityJan2004Dictionary.Namespace; } }
+            public override string TokenTypeUri { get { return _valueTypeUris[0]; } }
+            protected override string ValueTypeUri { get { return _valueTypeUris[0]; } }
+            public override bool SupportsTokenTypeUri(string tokenTypeUri)
+            {
+                for (int i = 0; i < _valueTypeUris.GetLength(0); ++i)
+                {
+                    if (_valueTypeUris[i] == tokenTypeUri)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        private class GenericXmlTokenEntry : TokenEntry
+        {
+            protected override XmlDictionaryString LocalName { get { return null; } }
+            protected override XmlDictionaryString NamespaceUri { get { return null; } }
+            protected override Type[] GetTokenTypesCore() { return new Type[] { typeof(GenericXmlSecurityToken) }; }
+            public override string TokenTypeUri { get { return null; } }
+            protected override string ValueTypeUri { get { return null; } }
+        }
+
+        private class UserNamePasswordTokenEntry : TokenEntry
+        {
+            protected override XmlDictionaryString LocalName { get { return CoreWCF.XD.SecurityJan2004Dictionary.UserNameTokenElement; } }
+            protected override XmlDictionaryString NamespaceUri { get { return CoreWCF.XD.SecurityJan2004Dictionary.Namespace; } }
+            protected override Type[] GetTokenTypesCore() { return new Type[] { typeof(UserNameSecurityToken) }; }
+            public override string TokenTypeUri { get { return SecurityJan2004Strings.UPTokenType; } }
+            protected override string ValueTypeUri { get { return null; } }
+        }
+
+        //class KerberosTokenEntry : BinaryTokenEntry
```

Comment:
Are you going to need this for the Windows Auth authentication you've worked on?

Reply:
yes...